### PR TITLE
Cknieling/texture cleanup bug

### PR DESCRIPTION
**Pull request by chr11115**
Fix a bug, where materials are not cleaned, when geometry is degenerate.

### CODE REVIEW

**Review by mistafunk — CHANGES_REQUESTED** [review]
Fix looks good. The change highlighted that the `initialize` and `compute` functions of Arnold and Stingray material nodes are mostly identical. I suggest we use this opportunity to introduce an abstract base class `SerlioMaterialNode` to get rid of the duplicate code.

**Comment by chr11115** [comment]
Introduced abstract base class MaterialNode (we never used serlio as prefix for material files (e.g. MaterialCommand, MaterialInfo, MaterialUtils..), except for the main plugin file serlioPlugin.cpp/h)

File: src/serlio/materials/ArnoldMaterialNode.h [hunk 1]
```diff
@@ -19,24 +19,25 @@
...
+	static MObject mOutMesh;
+
+private:
+	void declareMaterialStrings(MELScriptBuilder& sb);
+	void appendToMaterialScriptBuilder(MELScriptBuilder& sb, const MaterialInfo& matInfo,
+	                                   const std::wstring& shaderBaseName, const std::wstring& shadingEngineName);
+	std::wstring getBaseName();
+	MObject getInMesh();
+	MObject getOutMesh();
+	std::vector<std::string> getPluginDependencies();
```
Comment (mistafunk) [inline src/serlio/materials/ArnoldMaterialNode.h:42]: use `override` and `const` keywords for these virtual functions

File: src/serlio/materials/StingrayMaterialNode.h [hunk 2]
```diff
@@ -19,24 +19,25 @@
...
+	static MObject mOutMesh;
+
+private:
+	void declareMaterialStrings(MELScriptBuilder& sb);
+	void appendToMaterialScriptBuilder(MELScriptBuilder& sb, const MaterialInfo& matInfo,
+	                                   const std::wstring& shaderBaseName, const std::wstring& shadingEngineName);
+	std::wstring getBaseName();
+	MObject getInMesh();
+	MObject getOutMesh();
+	std::vector<std::string> getPluginDependencies();
```
Comment (mistafunk) [inline src/serlio/materials/StingrayMaterialNode.h:42]: also here, use `override` and `const` keywords for these virtual functions

File: src/serlio/materials/MaterialUtils.h [hunk 3]
```diff
@@ -32,4 +32,6 @@ std::wstring synchronouslyCreateShadingEngine(const std::wstring& desiredShading
 std::filesystem::path getStingrayShaderPath();
 
 bool textureHasAlphaChannel(std::wstring path);
```
Comment (mistafunk) [inline src/serlio/materials/MaterialUtils.h:34]: should use the opportunity to make this argument a const ref

File: src/serlio/materials/MaterialUtils.h [hunk 4]
```diff
@@ -32,4 +32,6 @@ std::wstring synchronouslyCreateShadingEngine(const std::wstring& desiredShading
...
+
+void resetMaterial(std::wstring meshName);
```
Comment (mistafunk) [inline src/serlio/materials/MaterialUtils.h:36]: same here, can use const ref argument

**Review by mistafunk — CHANGES_REQUESTED** [review]
* See inline comments...
* The free function `appendToMaterialScriptBuilder` is now redundant and not used, please remove it

File: src/serlio/materials/StingrayMaterialNode.h (hunk 2 above)
Comment (chr11115) [reply src/serlio/materials/StingrayMaterialNode.h:42]: done

File: src/serlio/materials/ArnoldMaterialNode.h (hunk 1 above)
Comment (chr11115) [reply src/serlio/materials/ArnoldMaterialNode.h:42]: done

File: src/serlio/materials/MaterialUtils.h (hunk 3 above)
Comment (chr11115) [reply src/serlio/materials/MaterialUtils.h:34]: done

File: src/serlio/materials/MaterialUtils.h (hunk 4 above)
Comment (chr11115) [reply src/serlio/materials/MaterialUtils.h:36]: done

**Review by mistafunk — CHANGES_REQUESTED** [review]
Still seeing that unused, redundant `appendToMaterialScriptBuilder` in StingrayMaterialNode anon namespace?!

**Comment by chr11115** [comment]
Now fixed.

**Review by mistafunk — APPROVED** [review]
LGTM

**Pull request merged**